b2sdk/replication/monitoring.py

```diff
@@ -0,0 +1,281 @@
+
+    # source & destination relation attrs
+    metadata_differs: Optional[bool] = None
+
+    LARGE_METADATA_SIZE: ClassVar[int] = 2048
+
+    @classmethod
+    def from_files(
+        cls, source_file: Optional[B2Path], destination_file: Optional[B2Path]
+    ) -> 'ReplicationAttrs':
+        params = {}
+
+        if source_file:
+            source_file_version = source_file.selected_version
+            params.update(
+                {
+                    'source_replication_status':
+                        source_file_version.replication_status,
+                    'source_has_hide_marker':
+                        source_file.is_visible(),
+                    'source_has_sse_c_enabled':
+                        source_file_version.server_side_encryption.mode == EncryptionMode.SSE_C,
+                    'source_has_large_metadata':
+                        source_file_version.headers_size >= cls.LARGE_METADATA_SIZE,
+                    'source_has_file_retention':
+                        source_file_version.file_retention is not NO_RETENTION_FILE_SETTING,
+                    'source_has_legal_hold':
+                        source_file_version.legal_hold is not LegalHold.UNSET,
+                }
+            )
+
+        if destination_file:
+            params.update(
+                {
+                    'destination_replication_status':
+                        destination_file.selected_version.replication_status,
+                }
+            )
+
+        if source_file and destination_file:
+            params.update(
+                {
+                    'metadata_differs':
+                        source_file.selected_version.file_info !=
+                        destination_file.selected_version.file_info,
+                }
+            )
+
+        return cls(**params)
+
+
+@dataclass
+class AbstractReplicationReport(metaclass=ABCMeta):
+    """
+    Aggregation of valuable information about file replication
+    after scanning source and (optionally) destination folders.
+    """
+
+    @abstractmethod
+    def add(self, source_file: Optional[B2Path], destination_file: Optional[B2Path]):
+        pass
+
+
+@dataclass
+class CountAndSampleReplicationReport(AbstractReplicationReport):
+    """
+    Replication report which groups and counts files by their `ReplicationAttrs` and
+    also stores first and last seen examples of such files.
+    """
+    counter_by_status: Counter[ReplicationAttrs] = field(default_factory=Counter)
+    samples_by_status_first: Dict[ReplicationAttrs, Tuple[FileVersion, FileVersion]] = field(
+        default_factory=dict
+    )
+    samples_by_status_last: Dict[ReplicationAttrs, Tuple[FileVersion, FileVersion]] = field(
+        default_factory=dict
+    )
+
+    def add(self, source_file: Optional[B2Path], destination_file: Optional[B2Path] = None):
+        status = ReplicationAttrs.from_files(source_file, destination_file)
```

Comment:
instead of using ReplicationAttrs directly here, assign it to a class field and use that, so that if someone wants to chance the ReplicationAttrs to something else they don't have to copy this entire method but 

```
class MyCountReport(CountAndSampleReplicationReport):
    ATTRS_CLASS = MyAttrsClass
```